UnityProject/Assets/Scripts/Objects/Engineering/Reflector.cs

```diff
@@ -0,0 +1,391 @@
+	{
+		[SerializeField]
+		private ReflectorType startingState = ReflectorType.Base;
+		private ReflectorType currentState = ReflectorType.Base;
+
+		[SerializeField]
+		private float startingAngle = 0;
+
+		[SerializeField]
+		private bool startSetUp;
+
+		[SerializeField]
+		private Transform spriteTransform;
+
+		private SpriteHandler spriteHandler;
+		private ObjectBehaviour objectBehaviour;
+		private RegisterTile registerTile;
+		private ObjectAttributes objectAttributes;
+		private Integrity integrity;
+
+		[SerializeField]
+		private int glassNeeded = 5;
+		[SerializeField]
+		private int reinforcedGlassNeeded = 10;
+		[SerializeField]
+		private int diamondsNeeded = 1;
+
+		[SerializeField]
+		//Use to check whether a bullet is a laser
+		private LayerMaskData laserData = null;
+
+		private bool isWelded;
+
+		[SyncVar(hook = nameof(SyncRotation))]
+		private float rotation;
+
+		#region LifeCycle
+
+		private void Awake()
+		{
+			spriteHandler = GetComponentInChildren<SpriteHandler>();
+			objectBehaviour = GetComponent<ObjectBehaviour>();
+			registerTile = GetComponent<RegisterTile>();
+			objectAttributes = GetComponent<ObjectAttributes>();
+			integrity = GetComponent<Integrity>();
+		}
+
+		private void OnValidate()
+		{
+			if (Application.isPlaying) return;
+
+			spriteHandler = GetComponentInChildren<SpriteHandler>();
+			currentState = startingState;
+			spriteHandler.ChangeSprite((int)startingState);
+			rotation = -startingAngle;
+			transform.localEulerAngles = new Vector3(0, 0, rotation);
+			spriteTransform.localEulerAngles = Vector3.zero;
+		}
+
+		private void Start()
+		{
+			if (CustomNetworkManager.IsServer == false) return;
+
+			ChangeState(startingState);
+			rotation = -startingAngle;
+			transform.localEulerAngles = new Vector3(0, 0, rotation);
+			spriteTransform.localEulerAngles = Vector3.zero;
+
+			if (startSetUp)
+			{
+				isWelded = true;
+				objectBehaviour.ServerSetPushable(false);
+			}
+		}
+
+		private void SyncRotation(float oldVar, float newVar)
+		{
+			rotation = newVar;
+			transform.localEulerAngles = new Vector3(0, 0, rotation);
+			spriteTransform.localEulerAngles = Vector3.zero;
+		}
+
+		private void OnEnable()
+		{
+			integrity.OnWillDestroyServer.AddListener(OnDestruction);
+		}
+
+		private void OnDisable()
+		{
+			integrity.OnWillDestroyServer.RemoveListener(OnDestruction);
+		}
+
+		#endregion
+
+		private void ChangeState(ReflectorType newState)
+		{
+			currentState = newState;
+			spriteHandler.ChangeSprite((int)newState);
+			objectAttributes.ServerSetArticleName(newState + " Reflector");
+		}
+
+		private void OnDestruction(DestructionInfo info)
+		{
+			DownGradeState(true);
+		}
+
+		#region HandApply
+
+		public bool WillInteract(HandApply interaction, NetworkSide side)
+		{
+			if (DefaultWillInteract.HandApply(interaction, side) == false) return false;
+
+			if (Validations.HasItemTrait(interaction.HandObject, CommonTraits.Instance.Welder)) return true;
```

Comment:
shouldn't use ``||`` here? Any check under the welder won't happen if the object doesn't have the welder trait.

Reply:
If it doesnt have the welder trait then it'll go to the next If statement, it only returns true if it is a welder. Does remind me though that i forgot to add a check for the welder to be on

Reply:
ah yeah, you're not returning if it doesn't have it.